### PR TITLE
TFLu: Support operator only input tensors.

### PR DESCRIPTION
**Pull request by mansnils**
Ethos-u relies on differentiating between having operator input tensors
as subgraph inputs or not.

### CODE REVIEW

**Comment by gbaned** [comment]
Thanks for contributing to TensorFlow Lite Micro.

To keep this process moving along, we'd like to make sure that you have completed the items on this list:
Read the [contributing guidelines for TensorFlow Lite Micro](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/CONTRIBUTING.md)
Created a [TF Lite Micro Github issue](https://github.com/tensorflow/tensorflow/issues/new?labels=comp%3Amicro&template=70-tflite-micro-issue.md)
Linked to the issue from the PR description

We would like to have a discussion on the Github issue first to determine the best path forward, and then proceed to the PR review.

**Comment by wangtz** [comment]
Hi Mans,

Trying to understand the context on this change. 

  tflite::testing::NodeConnection node_list[2] = {
      {
        {t0, t1, t2},  // t0: input (actual input part of subgraph inputs as well as
                       // operator inputs)
                       // t1: scratch1 (only in operator inputs)
                       // t2: scratch2 (only in operator inputs)
        {t3}           // output
      },
      {
        {t3},  // input
        {t4}   // output
      },
  };

Could you help me understand what's the behavior different with/without this change.

The online planning could handle this case pretty well.

**Comment by mansnils** [comment]
@wangtz  Ethos-U is depending on the offline planner and is using one or two additional operator input tensors (similar to scratch tensors, but part of the tflite file). We want these to remain as input tensors to the operator, but not as inputs to the subgraph.
Without this change the scratch tensors would be treated as read only and get incorrect address in the Ethos-u kernel. So the unit-test tests that and the model described by the node_list, is trying to picture a typical Ethos-U optimized model.

File: tensorflow/lite/micro/test_helpers.cc [hunk 1]
```diff
@@ -193,10 +195,17 @@ const Model* ModelBuilder::BuildModel(
 
   // TFLM only supports single subgraph.
   constexpr size_t subgraphs_size = 1;
+
+  // Find out number of subgraph inputs.
+  int num_tensors_in_subgraph_inputs = inputs.size(); // Default case = all inputs
+  if (num_subgraph_inputs > 0 && num_subgraph_inputs < inputs.size()) {
```
Comment (advaitjain) [inline tensorflow/lite/micro/test_helpers.cc:201]: I'd prefer having the separation between the default case and the ethos-u case be more explicit, and avoiding the need for `num_tensors_in_subgraph_inputs`.

Something like:

```cpp

if (num_subgraph_inputs == 0) {
  // this is the default case (i.e. for all platforms other than ethos-u)
  num_subgraph_inputs = inputs.size()
} else {
  // for ethos-u, a non-zero value of num_subgraph_inputs means that some of the inputs point to scratch buffers.   
  TF_LITE_DCHECK(num_subgraph_inputs < inputs.size());
}
```
and then this line can be:
```
 builder_->CreateVector(inputs.begin(), num_subgraph_inputs)
```

File: tensorflow/lite/micro/test_helpers.h [hunk 2]
```diff
@@ -91,7 +91,8 @@ const Model* GetSimpleModelWithBranch();
 const Model* GetModelWithOfflinePlanning(int num_tensors,
                                          const int32_t* metadata_buffer,
                                          NodeConnection* node_conn,
-                                         int num_conns);
+                                         int num_conns,
+                                         int num_subgraph_inputs = 0/*0 means all*/);
```
Comment (advaitjain) [inline tensorflow/lite/micro/test_helpers.h:95]: We haven't historically done a very good job documenting our functions but are trying to change that incrementally.

If you could document the function parameters, that would be excellent.

https://google.github.io/styleguide/cppguide.html#Function_Comments

**Review by advaitjain — CHANGES_REQUESTED** [review]
Apologies for the delay.

Mostly nits asking for documentation and then we'll be ready to get this merged.

File: tensorflow/lite/micro/micro_allocator.cc [hunk 3]
```diff
@@ -242,6 +242,18 @@ TfLiteStatus AllocationInfoBuilder::AddTensors(const SubGraph* subgraph,
     for (size_t n = 0; n < op->inputs()->size(); ++n) {
       const int tensor_index = op->inputs()->Get(n);
       AllocationInfo* current = &info_[tensor_index];
+
+      // In case operator input are not in subgraph inputs initialize them.
```
[the remark this answers is not in the document]
Comment (mansnils) [reply tensorflow/lite/micro/micro_allocator.cc:246]: Done

File: tensorflow/lite/micro/test_helpers.cc [hunk 4]
```diff
@@ -108,7 +108,8 @@ class ModelBuilder {
   // Constructs the flatbuffer model using `builder_` and return a pointer to
   // it. The returned model has the same lifetime as `builder_`.
   const Model* BuildModel(std::initializer_list<Tensor> inputs,
-                          std::initializer_list<Tensor> outputs);
+                          std::initializer_list<Tensor> outputs,
+                          size_t num_subgraph_inputs = 0);
```
[the remark this answers is not in the document]
Comment (mansnils) [reply tensorflow/lite/micro/test_helpers.cc:112]: Done

File: tensorflow/lite/micro/test_helpers.cc (hunk 1 above)
Comment (mansnils) [reply tensorflow/lite/micro/test_helpers.cc:201]: Done

File: tensorflow/lite/micro/test_helpers.h (hunk 2 above)
Comment (mansnils) [reply tensorflow/lite/micro/test_helpers.h:95]: Done

**Comment by mansnils** [comment]
@advaitjain Thanks for the comments. We believe this is not specific for Ethos-U, since we think potentially some other platform/product could use this as well.

**Review by advaitjain — APPROVED** [review]

**Pull request merged**